contracts/cosmwasm-vm/cw-ibc-core/src/ics04_channel/packet/receive_packet.rs

```diff
@@ -238,6 +232,92 @@ impl<'a> CwIbcCoreContext<'a> {
         Ok(())
     }
 
+    pub fn validate_packet_not_expired(
+        &self,
+        packet: &RawPacket,
+        host_height: Height,
+        host_timestamp: Timestamp,
+    ) -> Result<(), ContractError> {
+        let packet_height = packet
+            .timeout_height
+            .clone()
+            .map(|h| h.revision_height)
+            .unwrap_or(0);
+        if packet_height == 0 && packet.timeout_timestamp == 0 {
+            return Err(ContractError::IbcPacketError {
+                error: PacketError::InvalidTimeoutHeight,
+            });
+        }
+
+        if packet_height > 0 {
+            let packet_timeout_height = to_ibc_timeout_height(packet.timeout_height.clone())?;
+
+            if packet_timeout_height.has_expired(host_height) {
+                return Err(PacketError::LowPacketHeight {
+                    chain_height: host_height,
+                    timeout_height: packet_timeout_height,
+                })
+                .map_err(Into::<ContractError>::into)?;
+            }
+        }
+        if packet.timeout_timestamp > 0 {
+            let packet_timestamp = to_ibc_timestamp(packet.timeout_timestamp)?;
+
+            if let Expiry::Expired = host_timestamp.check_expiry(&packet_timestamp) {
+                return Err(ContractError::IbcPacketError {
+                    error: PacketError::LowPacketTimestamp,
+                });
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn validate_packet_expired(
+        &self,
+        packet: &RawPacket,
+        host_height: Height,
+        host_timestamp: Timestamp,
+    ) -> Result<(), ContractError> {
+        let packet_height = packet
+            .timeout_height
+            .clone()
+            .map(|h| h.revision_height)
+            .unwrap_or(0);
+        if packet_height == 0 && packet.timeout_timestamp == 0 {
+            return Err(ContractError::IbcPacketError {
+                error: PacketError::InvalidTimeoutHeight,
+            });
+        }
+
+        if packet_height > 0 {
+            let packet_timeout_height = to_ibc_timeout_height(packet.timeout_height.clone())?;
+
+            if !packet_timeout_height.has_expired(host_height) {
+                return Err(ContractError::IbcPacketError {
+                    error: PacketError::PacketTimeoutHeightNotReached {
+                        timeout_height: packet_timeout_height,
+                        chain_height: host_height,
+                    },
+                });
+            }
+        }
+        if packet.timeout_timestamp > 0 {
+            let packet_timestamp = to_ibc_timestamp(packet.timeout_timestamp)?;
```

Comment:
What if timestamp is expired but height is not

Reply:
I had the same question but it should be considered expired if any on param is expired right?

Reply:
hmm ... this also needs atleast one to be expired so may be i can just negate "validate packet not expired and map the errors". This will require both to be expired. Let me fix this.